### ISSUE TITLE
Expand the new replace image flow out to other blocks

### ISSUE THREAD

**Issue by kjellr** [WordPress/gutenberg#14795]
#14142 added an improved flow for replacing an image: 

![swap](https://user-images.githubusercontent.com/1202812/55480376-2900b380-55ee-11e9-82d3-9dba94d6fc0c.gif)

As per @mapk's [note in the thread](https://github.com/WordPress/gutenberg/pull/14142#issuecomment-476871253), this flow should be rolled out to other areas of the UI. Here's a quick list: 

- [ ] Video
- [ ] Audio
- [ ] Embeds* (Fixing this will also fix #11439)

In the following blocks, the edit button in the toolbar opens the media modal directly, so they just need the pencil icon swapped out for the new "Replace" icon: 

- [ ] Media & Text 
- [ ] Cover
- [ ] File

---

_*In the case of the Embed blocks (and maybe Video + Audio?) showing the preview in the placeholder may be too complicated. If that's the case, it may be fine to leave that part out._

**Comment by mapk** [comment]
I agree with your evaluation there. Let's remove the "Needs design feedback" and move toward a PR.

**Comment by talldan** [comment]
There's also https://github.com/WordPress/gutenberg/issues/14788 for the Media & Text block. @draganescu might have already started.

**Comment by draganescu** [comment]
Hi everyone, I have indeed already started! :) Will be posting a PR soon!

**Comment by draganescu** [comment]
I have added a WIP PR that implements the new flow to video, audio, media, cover, file. I'll check embeds asap :)

**Comment by draganescu** [comment]
Closed the #14918 PR and opened six instead:

- adds new edit flow to the file block #15515 
- adds new edit flow to the video block #15516 
- adds new edit flow to the media+text block #15517 
- adds new edit flow for the embed block #15518 
- adds new edit flow for the cover block #15519 
- adds new edit flow for the audio block #15520

**Comment by paaljoachim** [comment]
The edit icon is then being replaced in a lot of blocks that deal with images or files.
The new icon shows the swapping/changing of a file or image.

I know of one more that might benefit from the swap icon and that is the edit gallery (pencil) icon.
In general in a gallery one might want to change an image with another image. With the RC just released today one can click the image in the gallery and click the x to remove the image or click edit icon to through the media module remove an image.

That means the main reason to click the edit gallery would be to add a new image.
If most/all other blocks that deal with images/files have the icon replaced one might as well keep the consistency with the gallery block. As they all have to do with making a change to the existing image or file.

Or we just leave it as it is and see if anyone asks about this at a later point.

**Comment by afercia** [comment]
With regards to the problem that the "swap" (Edit) button disappears (focus loss) and may re-appear in a different position, see:
https://github.com/WordPress/gutenberg/issues/16048
https://github.com/WordPress/gutenberg/pull/15518#pullrequestreview-242319410

The first GIF posted by @kjellr makes sense to me and would allow to solve the problem:
- the button stays
- it works like a "toggle" to switch between preview and edit mode
- it communicates its "pressed" state

Seems to me this should apply to all the blocks that switch between the preview and the placeholder.

I quickly tried it on the current Audio block on master by:
- always rendering `BlockControls` (that means both when `state.editing` is false or true)
- making the button work as a toggle, with a `switchToView` callback, similar to `switchToEditing`
- adding to the button a `is-active` CSS class and `aria-pressed="true"` when editing, like other buttons that have a "pressed" state

Quickly testing that, seems to me:
- the focus loss doesn't happen
- the button stays in its position, solving the problem triggered by [the Fill/Slot thing](https://github.com/WordPress/gutenberg/pull/15518#issuecomment-497660420)
- the UI better communicates its state

Would this help to solve the issue and improve consistency? /Cc @draganescu 

Quick GIF from current master (with the old icon):

![audio swap](https://user-images.githubusercontent.com/1682452/59146295-112e1d00-89ed-11e9-966b-5956f3dc2b13.gif)

**Comment by draganescu** [comment]
@afercia I have implemented your idea in the six media blocks PRs and it worked like a charm. Will create a separate PR with the refactoring for the image block too. Great thinking! Thanks :)

**Comment by draganescu** [comment]
Hey @karmatosed, here it goes: 

- the initial exploration was ok,  so I started migrating it to other blocks. This issue seemed to be "close-able".
- First I opened one PR per block.
- Then closed them because we found that there was a better way to make one component to handle the editing flow in all media components.
- Then  @mtias argued that the icon didn't  solve the problem, so we moved towards having a text button.
- Then we stepped back from having one component that would wrap all media components to - having one edit button component.
- Then @mtias and @youknowriad suggested we should not have it in other components but change one media component first.
- Now I am 99.99% done with the whole thing in this PR #16200 which only needs someone to approve it, and it was very hard for me to get that to happen. Design and interaction wise @mapk and @jasmussen are happy so I hope this lands in master asap.

Now if this new image replacement flow works out OK, I will implement it in all the other blocks and only then this issue is completed :)

**Comment by karmatosed** [comment]
Ok that sounds great, it sort of sounds like we can close this issue is that right @draganescu?

**Comment by draganescu** [comment]
I started to move the media replace flow to the other blocks, since the final form of it is merged and appears to work allright.
